util/misc/ddg2dot.py

```diff
@@ -30,37 +44,71 @@
 for match in NODE_RE.finditer(text):
     num = match['number']
     name = match['name']
-    if name == 'artificial': # Prettify entry/exit names
+    if name == 'artificial':  # Prettify entry/exit names
         name = ['exit', 'entry'][match['other_name'] == '__optsched_entry']
 
     # Add the node to the graph. Include a node to make it clear what this is
     result.append(f'    n{num} [label="{name}:n{num}"];\n')
 
 result.append('\n')
 
+
+def create_edge_attrs(**attrs):
+    if not attrs:
+        return ''
+    attrtext = ' '.join(f'{key}="{value}"' for key, value in attrs.items())
+    return f' [{attrtext}]'
+
+
+def create_label(filtered_weights, weight, type_):
+    # The additional label text if we want to display the weight
+    # (that is, if the weight is not filtered out)
+    weight_label = '' if int(weight) in filtered_weights else ':' + weight
+    # The actual label text
+    return weight_label if type_ == 'data' else f'{type_}{weight_label}'
+
+
+def create_edge(from_, to, **attrs):
+    return f'    n{from_} -> n{to}{create_edge_attrs(**attrs)};\n'
+
+
+edges = set()
+
 # Create the edges in the graph
 for match in EDGE_RE.finditer(text):
     from_ = match['from']
     to = match['to']
     type_ = match['type']
     weight = match['weight']
 
-    # The additional label text if we want to display the weight
-    # (that is, if the weight is not filtered out)
-    weight_label = '' if int(weight) in filtered_weights else ':' + weight
-    # The actual label text
-    label = weight_label if type_ == 'data' else f'{type_}{weight_label}'
-    # The dot syntax to produce the actual label
-    label_section = f'label="{label}"'
+    result.append(
+        create_edge(
+            from_, to,
+            label=create_label(filtered_weights, weight, type_),
+        )
+    )
+    edges.add((from_, to))
+
+for match in EDGE_RE.finditer(basetext):
+    from_ = match['from']
+    to = match['to']
+    type_ = match['type']
+    weight = match['weight']
 
-    attr_section = f' [{label_section}]' if label else ''
-    # Add the edge
-    result.append(f'    n{from_} -> n{to}{attr_section};\n')
+    if (from_, to) not in edges:
+        result.append(
+            create_edge(
+                from_, to,
+                label=create_label(filtered_weights, weight, type_),
+                style="invis",
+            )
+        )
 
 # Graph is now finished:
 result.append('}\n')
 
 output = sys.stdout
-if args.output: output = open(args.output, 'w')
+if args.output:
+    output = open(args.output, 'w')
```

Comment:
The output file descriptor is not closed by the time the program ends.  This is a nit and the file will be closed anyway when the program exits, so I am okay with not doing anything about this comment.  But resource leaks are usually considered bad form.